src/hotspot/share/gc/shenandoah/shenandoahForwarding.inline.hpp

```diff
@@ -40,7 +40,7 @@ inline HeapWord* ShenandoahForwarding::get_forwardee_raw_unchecked(oop obj) {
   // On this path, we can encounter the "marked" object, but with NULL
   // fwdptr. That object is still not forwarded, and we need to return
   // the object itself.
-  markWord mark = obj->mark();
+  markWord mark = obj->mark_acquire();
```

Comment:
We also need the acquire barrier in fast path in generated code, right?

Reply:
Dang. I thought the beauty of self-fixing barriers is that we moved all fwdptr accesses to C++ (either GC or LRB), and all of them end up in this file. But there is `ShenandoahBarrierSetAssembler::cmpxchg_oop` that accesses the fwdptr directly. I shall see what can be done there.

Reply:
I believe only AArch64 needs a fix. x86_64 already has a strong semantics. See new commit.